mapbox_vector_tile/optimise.py

```diff
@@ -0,0 +1,272 @@
+        # in local tests seems to handle a few thousand lines without a
+        # problem.
+        min_dist = None
+        min_i = None
+        for i, line in enumerate(lines):
+            assert isinstance(line[0], MoveTo)
+            assert isinstance(line[-1], EndsAt)
+
+            moveto = line[0]
+
+            dist = abs(moveto.x - x) + abs(moveto.y - y)
+            if min_dist is None or dist < min_dist:
+                min_dist = dist
+                min_i = i
+
+        assert min_i is not None
+        line = lines.pop(min_i)
+        endsat = line[-1]
+        x = endsat.x
+        y = endsat.y
+        new_lines.append(line)
+
+    return new_lines
+
+
+def _rewrite_geometry(geom, new_lines):
+    """
+    Re-encode a list of Lines with absolute MoveTos as a continuous stream of
+    MVT geometry commands, each relative to the last. Replace geom with that
+    stream.
+    """
+
+    new_geom = []
+    x = 0
+    y = 0
+    for line in new_lines:
+        assert isinstance(line[0], MoveTo)
+        assert isinstance(line[-1], EndsAt)
+
+        moveto = line[0]
+        endsat = line[-1]
+
+        dx = moveto.x - x
+        dy = moveto.y - y
+        x = endsat.x
+        y = endsat.y
+
+        new_geom.append(9)  # move to, run_length = 1
+        new_geom.append(zigzag(dx))
+        new_geom.append(zigzag(dy))
+        new_geom.extend(line[1:-1])
+
+    # write the lines back out to geom
+    del geom[:]
+    geom.extend(new_geom)
+
+
+def optimise_multilinestring(geom):
+    # split the geometry into multiple lists, each starting with a move-to
+    # command and consisting otherwise of line-to commands. (perhaps with
+    # a close at the end? is that allowed for linestrings?)
+
+    lines = _decode_lines(geom)
+
+    # can't reorder anything unless it has multiple lines.
+    if len(lines > 1):
+        lines = _reorder_lines(lines)
+        _rewrite_geometry(geom, lines)
+
+
+def optimise_tile(tile_bytes):
+    """
+    Decode a sequence of bytes as an MVT tile and reorder the string table of
+    its layers and the order of its multilinestrings to save a few bytes.
+    """
+
+    t = tile()
+    t.ParseFromString(tile_bytes)
+
+    for layer in t.layers:
+        sto = StringTableOptimiser()
+
+        for feature in layer.features:
+            # (multi)linestrings only
+            if feature.type == 2:
+                optimise_multilinestring(feature.geometry)
+
+            sto.add_tags(feature.tags)
+
+        sto.update_string_table(layer)
+
+    return t.SerializeToString()
+
+
+if __name__ == 'main':
+    import argparse
+    import sys
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('input_file', help='Input MVT file')
+    parser.add_argument('--output-file', help='Output file, default is stdout')
+    args = parser.parse_args()
+
+    output_bytes = optimise_tile(open(args.input_file).read())
+
+    if args.output_file:
```

Comment:
Similarly, if you make `output_file` a `type=argparse.FileType('w')` and set `default=sys.stdout` it will handle this if/else for you.